sklearn/neighbors/base.py

```diff
@@ -334,18 +355,48 @@ class from an array representing our data set and ask who's
                     "or set algorithm='brute'" % self._fit_method)
             result = self._tree.query(X, n_neighbors,
                                       return_distance=return_distance)
-            return result
         else:
             raise ValueError("internal: _fit_method not recognized")
 
-    def kneighbors_graph(self, X, n_neighbors=None,
+        if not query_is_train:
+            return result
+
+        else:
+            if return_distance:
+                dist, neigh_ind = result
+            else:
+                neigh_ind = result
+
+            # If the query data is the same as the indexed data, we would like
+            # to ignore the first nearest neighbor of every sample, i.e
+            # the sample itself.
+            sample_mask = neigh_ind != sample_range
+
+            # Corner case: When the number of duplicates are more than the number
+            # of neighbors, the first NN will not be the sample, but a duplicate.
+            # In that case mask the first duplicate.
+            dup_gr_nbrs = np.all(sample_mask, axis=1)
+            sample_mask[:, 0][dup_gr_nbrs] = False
```

Comment:
I would probably have masked the last one but it is an arbitrary choice anyhow ^^